File: java/src/main/java/dojo/supermarket/model/Receipt.java
package dojo.supermarket.model;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class Receipt {

    private final List<ReceiptItem> items = new ArrayList<>();
    private final List<Discount> discounts = new ArrayList<>();

    public double getTotalPrice() {
        return items.stream().mapToDouble(ReceiptItem::totalPrice).sum() //
                + discounts.stream().mapToDouble(Discount::discountAmount).sum();
    }

    public void addProduct(Product p, double quantity, double price, double totalPrice) {
        items.add(new ReceiptItem(p, quantity, price, totalPrice));
    }

    public List<ReceiptItem> getItems() {
        return Collections.unmodifiableList(items);
    }

    public void addDiscount(Discount discount) {
        discounts.add(discount);
    }

    public List<Discount> getDiscounts() {
        return discounts;
    }
}

File: dart/supermarket_receipt/lib/supermarket/model/receipt.dart
import 'package:supermarket_receipt/supermarket/model/discount.dart';
import 'package:supermarket_receipt/supermarket/model/product.dart';
import 'package:supermarket_receipt/supermarket/model/receipt_item.dart';

class Receipt {
  
  final List<ReceiptItem> _items = [];
  final List<Discount> _discounts = [];

  void addProduct(Product p, double quantity, double price, double totalPrice) {
    _items.add(ReceiptItem(p, quantity, price, totalPrice));
  }

  void addDiscount(Discount discount) {
    _discounts.add(discount);
  }

  double getTotalPrice() {
    double total = 0.0;
    for (var item in _items) {
      total += item.getTotalPrice();
    }
    for (var discount in _discounts) {
      total += discount.getDiscountAmount();
    }
    return total;
  }

  List<ReceiptItem> getItems() => _items;

  List<Discount> getDiscounts() => _discounts;
}